README.md
# TimeAgoInWords

[![Gem Version][GV img]][Gem Version]
[![Build Status][BS img]][Build Status]
[![Dependency Status][DS img]][Dependency Status]
[![Code Climate][CC img]][Code Climate]
[![Coverage Status][CS img]][Coverage Status]

## Description

Humanize elapsed time from some Time instance to Time.now, e.g. '2 hours and 1 minute ago'

## Installation

`$ gem install time_ago_in_words` or add to your [Gemfile][] this line: `gem 'time_ago_in_words'` then run [bundle install][]

## Usage

Just `require 'time_ago_in_words'` and then call Time#ago_in_words method:

```ruby
require 'time_ago_in_words'

# Comparing from now:
(Time.now - 10).ago_in_words #=> "10 seconds ago"
(Time.now - 1).ago_in_words  #=> "1 second ago"
(Time.now - 60).ago_in_words #=> "1 minute ago"
(Time.now - 63).ago_in_words #=> "1 minute and 3 seconds ago"

# This is my current time so you can compare
Time.now #=> 2013-03-06 02:19:23 -0300

Time.local(1981,03,03,20,30,40).ago_in_words #=> "690 days and 5 hours ago"
Time.local(2013,03,03,20,30,40).ago_in_words #=> "2 days and 5 hours ago"
Time.local(2013,03,04,20,30,40).ago_in_words #=> "1 day and 5 hours ago"
Time.local(2013,03,05,20,30,40).ago_in_words #=> "5 hours and 48 minutes ago"
Time.local(2013,03,05,21,13,40).ago_in_words #=> "5 hours and 5 minutes ago"
Time.local(2013,03,06,00,30,40).ago_in_words #=> "1 hour and 48 minutes ago"
Time.local(2013,03,06,01,11,40).ago_in_words #=> "1 hour and 7 minutes ago"
Time.local(2013,03,06,01,27,40).ago_in_words #=> "51 minutes and 43 seconds ago"
Time.local(2013,03,06,02,19,20).ago_in_words #=> "3 seconds ago"
```

## Contributing

1. Fork it.
2. Make your feature addition or bug fix and create your feature branch.
3. Update the [Change Log][].
3. Add specs/tests for it. This is important so I don't break it in a future version unintentionally.
4. Commit, create a new Pull Request.
5. Check that your pull request passes the [build][travis pull requests].

### TODO
+ Add DateTime support. Currently only available for Time objects.
+ Add "N months ago"
+ Add "N years ago"
+ Add "N decades ago"
+ Add "N centuries ago"
+ Add more rspec Examples
+ Implement Time mocking for testing, instead of abusing Time.now on spec
+ Extract [time_ago_in_words][] and implement that alternative precision syntax, e.g. "less than a minute ago"

## License

Released under the MIT License. See the [LICENSE][] file for further details.

## Links

[RubyGems][] | [Documentation][] | [Source][] | [Bugtracker][] | [Build Status][] | [Dependency Status][] | [Code Climate][]


[bundle install]: http://gembundler.com/man/bundle-install.1.html
[Gemfile]: http://gembundler.com/man/gemfile.5.html
[LICENSE]: LICENSE.md
[Change Log]: CHANGELOG.md

[RubyGems]: https://rubygems.org/gems/time_ago_in_words
[Documentation]: http://rubydoc.info/gems/time_ago_in_words
[Source]: https://github.com/elgalu/time_ago_in_words
[Bugtracker]: https://github.com/elgalu/time_ago_in_words/issues

[travis pull requests]: https://travis-ci.org/elgalu/time_ago_in_words/pull_requests

[Gem Version]: https://rubygems.org/gems/time_ago_in_words
[Build Status]: https://travis-ci.org/elgalu/time_ago_in_words
[Dependency Status]: https://gemnasium.com/elgalu/time_ago_in_words
[Code Climate]: https://codeclimate.com/github/elgalu/time_ago_in_words
[Coverage Status]: https://coveralls.io/r/elgalu/time_ago_in_words

[GV img]: https://badge.fury.io/rb/time_ago_in_words.png
[BS img]: https://travis-ci.org/elgalu/time_ago_in_words.png
[DS img]: https://gemnasium.com/elgalu/time_ago_in_words.png
[CC img]: https://codeclimate.com/github/elgalu/time_ago_in_words.png
[CS img]: https://coveralls.io/repos/elgalu/time_ago_in_words/badge.png?branch=master

[time_ago_in_words]: http://apidock.com/rails/ActionView/Helpers/DateHelper/time_ago_in_words


CHANGELOG.md
## [In git](https://github.com/elgalu/time_ago_in_words/compare/v0.1.0...HEAD)

### New Features
* n/a

### Bugfixes
* n/a

### Chores
* n/a

## [v0.1.0](https://github.com/elgalu/time_ago_in_words/tree/v0.1.0)

### New Features
* n/a

### Bugfixes
* Removed support for ruby 1.8.7 due to 'time ouf of range' errors. (Leo Gallucci)

### Chores
* Applied Extract Method refactor on Time#ago_in_words(). (Leo Gallucci)

## [v0.0.1](https://github.com/elgalu/time_ago_in_words/tree/v0.0.1)

## First gem release

### Features:
* Humanize elapsed time from some Time instance to Time.now, e.g. '2 hours and 1 minute ago' (Leo Gallucci)


spec/spec_helper.rb
require 'simplecov'
require 'coveralls'

SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  SimpleCov::Formatter::HTMLFormatter,
  Coveralls::SimpleCov::Formatter
]
SimpleCov.start

require 'time_ago_in_words'

# Require this file using `require "spec_helper"` within each of your specs
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
  config.filter_run :focus

  # Run specs in random order to surface order dependencies.
  config.order = 'random'
end


spec/time_ago_in_words_spec.rb
require 'spec_helper'

describe TimeAgoInWords do
  it 'should have a version number' do
    TimeAgoInWords::VERSION.should_not be_nil
  end

  describe '#ago_in_words' do
    context 'invalid dates' do
      it "returns 'just now' when elapsed seconds are less than 1" do
        Time.now.ago_in_words.should == 'just now'
        (Time.now - 0.5).ago_in_words.should == 'just now'
        (Time.now + 0.5).ago_in_words.should == 'just now'
      end

      it "returns empty string on future dates" do
        (Time.now + 1.6).ago_in_words.should == ''
        (Time.now + 100).ago_in_words.should == ''
        (Time.now + 9999).ago_in_words.should == ''
      end

      it "returns 'a very very long time ago' when more than 300 years ago" do
        Time.local(1713,01,01).ago_in_words.should == 'a very very long time ago'
      end

      it "can handle many days ago" do
        (Time.now - 60*60*24*9999).ago_in_words.should == '9999 days ago'
      end

      it "returns N seconds ago, N seconds ago!" do
        (Time.now - 2).ago_in_words.should == '2 seconds ago'
        (Time.now - 33).ago_in_words.should == '33 seconds ago'
      end

      it "returns 1 second/minute/hour/day ago when 1 ago!" do
        (Time.now - 1).ago_in_words.should        == '1 second ago'
        (Time.now - 60).ago_in_words.should       == '1 minute ago'
        (Time.now - 60*60).ago_in_words.should    == '1 hour ago'
        (Time.now - 60*60*24).ago_in_words.should == '1 day ago'
      end

      it "returns 1 second/minute/hour/day and 1 second/minute/hour/day ago when appropriate" do
        (Time.now - 1 - 60).ago_in_words.should             == '1 minute and 1 second ago'
        (Time.now - 1*60*60 - 60).ago_in_words.should       == '1 hour and 1 minute ago'
        (Time.now - 1*60*60*24 - 60*60).ago_in_words.should == '1 day and 1 hour ago'
      end

      it "returns 1 second/minute/hour/day and 1 second/minute/hour/day ago ignoring third elements (residual irrelevant precisions)" do
        (Time.now - 1*60*60 - 60 - 10).ago_in_words.should        == '1 hour and 1 minute ago'
        (Time.now - 1*60*60*24 - 1*60*60 - 100).ago_in_words.should == '1 day and 1 hour ago'
        (Time.now - 9*60*60*24 - 3*60*60 - 100).ago_in_words.should == '9 days and 3 hours ago'
      end
    end
  end
end


lib/time_ago_in_words/methods.rb
module TimeAgoInWords
  # Convert elapsed seconds from Time.now to the current Time instance
  #   into an English human readable text like '4 hours and 1 minute ago'
  # It is not well suited for dates older than 1 year
  #
  # @return [String] the humanized elapsed time in pairs
  #   Always returns a pair of words like:
  #     '23 days and 3 hours ago'
  #     '5 hours and 44 minutes ago'
  #     ' 1 minute and 11 seconds ago'
  #
  #   Or just 1 when no hours/minutes/seconds available:
  #     '23 days ago'
  #     '17 hours ago'
  #     '1 minute ago'
  #     '1 second ago'
  #
  #   Ẃhen 0 seconds or Time instance is a future date, returns empty string:
  #     ""
  #
  # @note Inspired from http://stackoverflow.com/a/4136485/511069
  def ago_in_words
    return 'a very very long time ago' if self.year < 1800
    secs = Time.now - self
    return 'just now' if secs > -1 && secs < 1
    return '' if secs <= -1
    pair = ago_in_words_pair(secs)
    ary = ago_in_words_singularize(pair)
    ary.size == 0 ? '' : ary.join(' and ') << ' ago'
  end

  private

  # @private
  def ago_in_words_pair(secs)
    [[60, :seconds], [60, :minutes], [24, :hours], [100_000, :days]].map{ |count, name|
      if secs > 0
        secs, n = secs.divmod(count)
        "#{n.to_i} #{name}"
      end
    }.compact.reverse[0..1]
  end

  # @private
  def ago_in_words_singularize(pair)
    if pair.size == 1
      pair.map! {|part| part[0, 2].to_i == 1 ? part.chomp('s') : part }
    else
      pair.map! {|part| part[0, 2].to_i == 1 ? part.chomp('s') : part[0, 2].to_i == 0 ? nil : part }
    end
    pair.compact
  end
end
